scripts/zfs-probes.sh

```diff
@@ -0,0 +1,114 @@
+#!/bin/sh
+#
+# This script is designed to facilitate in-tree development and testing
+# by installing symlinks on your system which refer to in-tree helper
+# utilities.  These helper utilities must be installed to in order to
+# exercise all ZFS functionality.  By using symbolic links and keeping
+# the scripts in-tree during development they can be easily modified
+# and those changes tracked.
+#
+# Use the following configuration option to override the installation
+# paths for these scripts.  The correct path is automatically set for
+# most distributions but you can optionally set it for your environment.
+#
+#   --with-mounthelperdir=DIR  install mount.zfs in dir [/sbin]
+#   --with-udevdir=DIR         install udev helpers [default=check]
+#   --with-udevruledir=DIR     install udev rules [default=UDEVDIR/rules.d]
+#   --sysconfdir=DIR           install zfs configuration files [PREFIX/etc]
+#
+
+fail() {
+	echo "${PROG}: $1" >&2
+	exit 1
+}
+
+usage() {
+cat << EOF
+USAGE:
+$0 fp
+
+DESCRIPTION:
+	Generate dtrace probe functions or prototype macros for all
+	DTRACE_PROBE calls in the zfs source directory.
+
+OPTIONS:
+	-f	Generate functions
+	-p	Generate prototypes
+	-d directory
+
+EOF
+}
+GEN_FUNCS="no"
+GEN_PROTO="no"
+DIRECTORY=""
+
+while getopts 'fpd:' OPTION; do
+	case $OPTION in
+	h)
+		usage
+		exit 1
+		;;
+	f)
+		GEN_FUNCS="yes"
+		;;
+	p)
+		GEN_PROTO="yes"
+		;;
+	d)	DIRECTORY="$OPTARG"
+		;;
+	?)
+		usage
+		exit
+		;;
+	esac
+done
+
+if [ "$GEN_FUNCS" = "yes" ] && [ "$GEN_PROTO" = "yes" ]; then
+	fail "Specify -f or -p but not both"
+fi
+
+if [ "$GEN_FUNCS" = "no" ] && [ "$GEN_PROTO" = "no" ]; then
+	fail "Either -f or -p must be specified"
+fi
+
+if [ "$DIRECTORY" = "" ]; then
+	fail "A vaild directory must be specified"
+fi
+
+if [ "$GEN_FUNCS" = "yes" ]; then
+	SED_COMMAND='s/DTRACE_PROBE/FUNC_DTRACE_PROBE/'
+else
+	SED_COMMAND='s/DTRACE_PROBE/PROTO_DTRACE_PROBE/'
+fi
+
+cat ${DIRECTORY}/*.c | awk '
+BEGIN { OFS = ""}
+/DTRACE_PROBE/,/\);/ {
+	where = match ($0, "\\#define")
+	if (where) {
+		where = 0
```

Comment:
It would be great to have
1. a comment explaining what this is trying to do at a high level, which I think I understand:
for each `DTRACE_PROBE[1234](...)` that it finds, print out `FUNC_DTRACE_PROBE[1234](...)` or `PROTO_DTRACE_PROBE[1234](...)`.
2. comments explaining how we do this using this awk script, which is way over my head.  I'm guessing the complexity is to handle line breaks?

Reply:
I added comments and was able to simplify things by leveraging cpp.